internal/providers/vault/certificate.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+const (
+	vaultJwtAudience    = "vault.namespace.systems"
+	vaultRequestTimeout = 10 * time.Second
+)
+
+func Register() {
+	p := &provider{
+		vaultClients: make(map[vaultAuth]*vaultclient.Client),
+	}
+	combined.RegisterSecretsProvider(
+		func(ctx context.Context, srvRef *secrets.ServerRef, cfg *vault.Certificate) ([]byte, error) {
+			ca := cfg.GetCa()
+			if ca == nil {
+				return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing CA configuration")
+			}
+
+			vaultClient, err := p.Login(ctx, ca, vaultJwtAudience)
+			if err != nil {
+				return nil, err
+			}
+
+			commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(srvRef.RelPath, "/", "-"), cfg.BaseDomain)
+			return p.IssueCertificate(ctx, vaultClient, ca.GetIssuer(), commonName)
+		},
+	)
+}
+
+type vaultAuth struct {
+	addr       string
+	authMethod string
+}
+
+type tlsKeyPair struct {
+	PrivateKey  string   `json:"private_key"`
```

Comment:
One minor annoyance with these values is that it isn't immediately obvious whether the data is in PEM or DER format. I happen to remember that Vault returns them in PEM format, but there's nothing here that hints to that. It'd be nice if these fields would be more explicit, e.g. private_key_pem instead of private_key, or at least that should be documented somehow, maybe in a comment above the field.

Reply:
Changed, now each field has suffix `_pem`